src/sailfish-iptables.c

```diff
@@ -0,0 +1,1241 @@
+
+	str_rule = g_string_free(rule,FALSE);
+
+	if(rval == OK && str_rule && params->operation != UNDEFINED)
+	{
+		if(params->operation == ADD)
+		{	
+			if(!(result = connman_iptables_append(SAILFISH_IPTABLES_TABLE_NAME,
+				ipt_operation, str_rule)))
+				DBG("%s %s %s %s", PLUGIN_NAME, "connman_iptables_append",
+					ipt_operation, str_rule);
+			else
+				DBG("%s %s %s %s  %d", PLUGIN_NAME,
+					"connman_iptables_append failure", ipt_operation, str_rule,
+					result);
+		}
+		else if(params->operation == REMOVE)
+		{
+			if(!(result = connman_iptables_delete(SAILFISH_IPTABLES_TABLE_NAME, 
+				ipt_operation, str_rule)))
+				DBG("%s %s %s %s", PLUGIN_NAME,
+					"connman_iptables_delete success", ipt_operation, str_rule);
+			else
+				DBG("%s %s %s %s %d", PLUGIN_NAME,
+					"connman_iptables_delete failure", ipt_operation, str_rule,
+					result);
+		}
+	
+		if(!result)
+		{
+			if(!(result = connman_iptables_commit(SAILFISH_IPTABLES_TABLE_NAME)))
+				DBG("%s %s %d", PLUGIN_NAME, "connman_iptables_commit", result);
+			else
+			{
+				DBG("%s %s %d", PLUGIN_NAME,
+					"connman_iptables_commit failed:", result);
+					
+				if(params->operation == ADD)
+				{
+					if(connman_iptables_delete(SAILFISH_IPTABLES_TABLE_NAME, 
+						ipt_operation, str_rule))
+						ERR("Cannot revert rule (%s) - clear/restart connman",
+							str_rule);
+					else
+						DBG("connman_iptables_delete reverted %s", str_rule);
+				}
+				
+				rval = INVALID;
+			}
+		}
+		else
+		{
+			if(params->operation == REMOVE)
+				rval = RULE_DOES_NOT_EXIST;
+			else
+				rval = INVALID_REQUEST;
+		}
+	}
+	
+	g_free(str_rule);
+	
+	return params->operation == UNDEFINED ? INVALID_REQUEST : rval;
+}
+
+static api_result __allow_incoming(rule_params* params)
+{
+	return __add_rule_to_iptables(params, OPERATION_IN |OPERATION_ACCEPT);
+}
+
+static api_result __allow_outgoing(rule_params* params)
+{
+	return __add_rule_to_iptables(params, OPERATION_OUT | OPERATION_ACCEPT);
+}
+
+static api_result __deny_incoming(rule_params* params)
+{
+	return __add_rule_to_iptables(params,OPERATION_IN | OPERATION_DENY);
+}
+
+static api_result __deny_outgoing(rule_params* params)
+{
+	return __add_rule_to_iptables(params,OPERATION_OUT | OPERATION_DENY);
+}
+
+DBusMessage* __process_request(DBusMessage *message,
+	api_result (*func)(rule_params* params), rule_args args)
+{
+	api_result result = INVALID;
+	rule_params *params = NULL;
+	
+	if((params =  __get_parameters_from_message(message,args)))
+	{	
+		if((result = func(params)) == OK)
+		{
+			DBusMessage *signal = __signal_from_rule_params(params);
+			if(signal)
+				sailfish_iptables_dbus_send_signal(signal);
+		}
+		else
+			ERR("%s %s %d",
+				"__process_request():", "request was not successful",
+				result);
+	}
+	
+	rule_params_free(params);
+	
+	DBusMessage* reply = sailfish_iptables_dbus_method_return(message,
+			DBUS_TYPE_UINT16,	(guint16)result,
+			DBUS_TYPE_STRING, 	api_result_message(result),
+			DBUS_TYPE_INVALID);
```

Comment:
Unless I'm missing something, all of these are wrong...  should pass pointers, not values. (and should use dbus types, not glib ones)

Reply:
Fixed in 257c056893e8799c614ee7d78d2741dfb28bdc66